core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -292,16 +292,17 @@ protected boolean binaryArithmeticWithStrings(
     }
 
     RelDataType commonType;
-    if (SqlTypeUtil.sameNamedType(type1, type2)) {
+    // First check when one of the types is a string (see JavaDoc on cmmonTypeForBinaryComparison)
+    commonType = commonTypeForBinaryComparison(type1, type2);
```

Comment:
previous to this change, coercion would not insert casts when comparing say int to double, because they didn't have the same type name. that was a bug.

Reply:
agree